# Gather

In [122]:
import requests
import tweepy
import pandas as pd
import numpy as np
import json
import os
import matplotlib.pyplot as plt
%matplotlib inline 

import seaborn as sns
sns.set()

###  The WeRateDogs Twitter archive

In [123]:
df_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [124]:
# Making a list of tweet_id
tweet_ids = df_archive.tweet_id

### Downloading the Tweet Image Predictions Programmatically

In [37]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)
with open(os.path.join(url.split('/')[-1]), mode='wb') as file:
    file.write(response.content)

In [7]:
df_img = pd.read_csv('image-predictions.tsv', sep="\t")

### API for Additional Tweets Data

In [ ]:
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [ ]:
# Initialize empty list to store tweets: tweets_data
consumer_key = 
consumer_secret = 
access_token = 
access_secret = 

tweets_data = []
deleted = []

for tweet_id in tweet_ids:
    print(tweet_id)
    try:
        tweet = api.get_status(tweet_id)
        tweets_data.append(tweet)
    except:
        deleted.append(tweet)

In [ ]:
# Create empty list for each twitter status to be stored
json_strs = []

# Appending each string to the text file
for tweet_data in tweets_data:
    json_str = json.dumps(tweet_data._json)
    json_strs.append(json_str)

In [ ]:
# Writing text file from JSON
with open('tweet_json.txt', 'w') as outfile:
    json.dump(json_strs, outfile)

In [8]:
# Creating an object from text file
with open('tweet_json.txt', 'r') as fp:
    obj = json.load(fp)

In [9]:
# From the object of list to make a dictionary
tweets_dict = []
for line in obj:
    tweet_line = json.loads(line)
    tweets_dict.append(tweet_line)

In [308]:
# Create a pandas DataFrame using the keys from the first line
columns = ['created_at', 'id', 'id_str', 'text', 'truncated', 
           'entities', 'extended_entities', 'source', 
           'in_reply_to_status_id', 'in_reply_to_status_id_str', 
           'in_reply_to_user_id', 'in_reply_to_user_id_str', 
           'in_reply_to_screen_name', 'user', 'geo', 'coordinates', 
           'place', 'contributors', 'is_quote_status', 'retweet_count', 
           'favorite_count', 'favorited', 'retweeted', 
           'possibly_sensitive', 'possibly_sensitive_appealable', 'lang']

df_tweets = pd.DataFrame(tweets_dict, columns=columns)

# Assess

### The WeRateDogs Twitter archive: df_archive

In [309]:
df_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [310]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [311]:
df_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

<b>Investigating dog names</b>

In [312]:
df_archive.groupby('name')['name'].nunique()

name
Abby            1
Ace             1
Acro            1
Adele           1
Aiden           1
Aja             1
Akumi           1
Al              1
Albert          1
Albus           1
Aldrick         1
Alejandro       1
Alexander       1
Alexanderson    1
Alf             1
Alfie           1
Alfy            1
Alice           1
Amber           1
Ambrose         1
Amy             1
Amélie          1
Anakin          1
Andru           1
Andy            1
Angel           1
Anna            1
Anthony         1
Antony          1
Apollo          1
               ..
Ziva            1
Zoe             1
Zoey            1
Zooey           1
Zuzu            1
a               1
actually        1
all             1
an              1
by              1
getting         1
his             1
incredibly      1
infuriating     1
just            1
life            1
light           1
mad             1
my              1
not             1
officially      1
old             1
one             1
quite           1
space

<b>Investigating unusual numerator</b>

In [313]:
df_archive[df_archive.rating_numerator == 0]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                                                  source  \
315   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1016  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                         text  \
315         When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag   
1016  PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX   

      retweeted_status_id  retweeted_status_user_id  \
315                   NaN                       NaN   
1016                  NaN                       NaN   

     retweeted_status_timestamp  \
315                         NaN   
1016                        NaN   

                                                                                                                                                                                        expanded_urls  \
315   https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1   
1016                                                                                                                                  https://twitter.com/dog_rates/status/746906459439529985/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
315                  0                  10  None  None    None   None  None  
1016                 0                  10  None  None    None   None  None

<b>Investigating on unusual denominator</b>

In [314]:
df_archive.rating_denominator.unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2])

In [315]:
df_archive[(df_archive.rating_denominator == 0) |
          (df_archive.rating_denominator == 15) |
          (df_archive.rating_denominator == 70) |
          (df_archive.rating_denominator == 7) |
          (df_archive.rating_denominator == 11) |
          (df_archive.rating_denominator == 150) |
          (df_archive.rating_denominator == 170) |
          (df_archive.rating_denominator == 20) |
          (df_archive.rating_denominator == 50) |
          (df_archive.rating_denominator == 90) |
          (df_archive.rating_denominator == 80) |
          (df_archive.rating_denominator == 40) |
          (df_archive.rating_denominator == 130) |
          (df_archive.rating_denominator == 110) |
          (df_archive.rating_denominator == 16) |
          (df_archive.rating_denominator == 120) |
          (df_archive.rating_denominator == 2)]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                                                  source  \
313   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
342   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
433   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
516   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
784   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
902   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1228  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1254  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1274  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1433  <a href="http://twitter.com/download/iphone" rel="nofollow">

<b>Investigating on unusual numerator</b>

In [316]:
df_archive['numerator'] = df_archive.text.str.extract('(\d[.,]?\d)')
df_archive['denominator'] = df_archive.text.str.extract('(\/(?=.*?\d)\d*[.,]?\d)')

#strip '/' from denominator
df_archive['denominator'] = df_archive['denominator'].str.replace('/', '')
print(df_archive['denominator'].unique())
print(df_archive['numerator'].unique())

['10' '00' '15' '70' '7' '11' '150' '170' '20' '50' '90' '80' '40' '130'
 '110' '16' '120' '2']
['13' '12' '14' '17' '15' '11' '21' '10' '42' '66' '98' '23' '60' '18' '20'
 '96' '9.7' '00' '84' '24' '46' '16' '50' '2.0' '31' '47' '99' '80' '45'
 '40' '44' '33' '97' '61' '26' '9.5' '92' '88' '85' '22' '8.9' '19']


/Users/emimimura/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.
/Users/emimimura/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  


Finding out the correct numbers are, then dropping the new columns I created since I'm done using them.

In [317]:
df_archive[(df_archive['numerator'] == '9.7') |
          (df_archive['numerator'] == '00') |
          (df_archive['numerator'] == '2.0') |
          (df_archive['numerator'] == '9.5') |
          (df_archive['numerator'] == '8.9')]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
340   832215909146226688                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
695   786709082849828864                    NaN                  NaN   
1018  746818907684614144           6.914169e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
2347  666057090499244032                    NaN                  NaN   

                      timestamp  \
340   2017-02-16 13:11:49 +0000   
387   2017-02-01 01:11:25 +0000   
695   2016-10-13 23:23:56 +0000   
1018  2016-06-25 21:34:37 +0000   
1689  2015-12-28 05:07:27 +0000   
2347  2015-11-16 00:55:59 +0000   

                                                                                  source  \
340   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
387   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
695   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1018  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1689  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2347  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
340   RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…   
387                                                                          I was going to do 007/10, but the joke wasn't worth the &lt;10 rating   
695               This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   
1018                                                                               Guys... Dog Jesus 2.0\n13/10 buoyant af https://t.co/CuNA7OwfKQ   
1689                                    I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace   
2347                  My oh my. This is a rare blond Canadian terrier on wheels. Only $8.98. Rather docile. 9/10 very rare https://t.co/yWBqbrzy8O   

      retweeted_status_id  retweeted_status_user_id  \
340          7.867091e+17              4.196984e+09   
387                   NaN                       NaN   
695                   NaN                       NaN   
1018                  NaN                       NaN   
1689                  NaN                       NaN   
2347                  NaN                       NaN   

     retweeted_status_timestamp  \
340   2016-10-13 23:23:56 +0000   
387                         NaN   
695                         NaN   
1018                        NaN   
1689                        NaN   
2347                        NaN   

                                                        expanded_urls  \
340   https://twitter.com/dog_rates/status/786709082849828864/photo/1   
387                                                               NaN   
695   https://twitter.com/dog_rates/status/786709082849828864/photo/1   
1018  https://twitter.com/dog_rates/status/746818907684614144/photo/1   
1689                                                              NaN   
2347  https://twitter.com/dog_rates/status/666057090499244032/photo/1   

      rating_numerator  rating_denominator   name doggo floofer pupper puppo  \
340                 75                  10  Logan  None    None   None  None   
387                  7                  10   None  None    None   None  None   
695                 75                  10  Logan  None    None   None  None   
1018                13                  10   None  None    None   None  None   
1689                 5   

In [318]:
df_archive = df_archive.drop(['denominator', 'numerator'], axis=1)

<b>Investigating the stage names</b>

In [320]:
df_archive[(df_archive.doggo == "doggo") & (df_archive.floofer == "floofer")]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
200  854010172552949760                    NaN                  NaN   

                     timestamp  \
200  2017-04-17 16:34:26 +0000   

                                                                                 source  \
200  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                     text  \
200  At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
200                  NaN                       NaN                        NaN   

                                                                                                                       expanded_urls  \
200  https://twitter.com/dog_rates/status/854010172552949760/photo/1,https://twitter.com/dog_rates/status/854010172552949760/photo/1   

     rating_numerator  rating_denominator  name  doggo  floofer pupper puppo  
200                11                  10  None  doggo  floofer   None  None

In [321]:
df_archive[(df_archive.doggo == "doggo") & (df_archive.pupper == "pupper")]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
460   817777686764523521                    NaN                  NaN   
531   808106460588765185                    NaN                  NaN   
565   802265048156610565           7.331095e+17         4.196984e+09   
575   801115127852503040                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
733   781308096455073793                    NaN                  NaN   
778   775898661951791106                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1113  733109485275860992                    NaN                  NaN   

                      timestamp  \
460   2017-01-07 16:59:28 +0000   
531   2016-12-12 00:29:28 +0000   
565   2016-11-25 21:37:47 +0000   
575   2016-11-22 17:28:25 +0000   
705   2016-10-11 00:34:48 +0000   
733   2016-09-29 01:42:20 +0000   
778   2016-09-14 03:27:11 +0000   
822   2016-08-29 03:00:36 +0000   
889   2016-07-31 16:50:42 +0000   
956   2016-07-09 01:08:47 +0000   
1063  2016-06-10 00:39:48 +0000   
1113  2016-05-19 01:38:16 +0000   

                                                                                  source  \
460   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
531   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
565   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
575   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
705   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
733                      <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   
778   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
822   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
889   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
956   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1063  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1113  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                      text  \
460   This is Dido. She's playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7   
531                       Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   
565                                                                                                  Like doggo, like pupper version 2. Both 11/10 https://t.co/9IxWAXFqze   
575                       This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj   
705   This is Pinot. He's a sophisticated doggo. You can tell by the hat. Also pointier than your average pupper. Still 10/10 would pet cautiously https://t.co/f2wmLZTPHd   
733                                                                                                             Pupper butt 1, Doggo 0. Both 12/10 https://t.co/WQvcPEpH2u   
778                                                                              RT @dog_rates: Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda   
822                               

In [322]:
df_archive[(df_archive.doggo == "doggo") & (df_archive.puppo == "puppo")]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
191  855851453814013952                    NaN                  NaN   

                     timestamp  \
191  2017-04-22 18:31:02 +0000   

                                                                                 source  \
191  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               text  \
191  Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
191                  NaN                       NaN                        NaN   

                                                       expanded_urls  \
191  https://twitter.com/dog_rates/status/855851453814013952/photo/1   

     rating_numerator  rating_denominator  name  doggo floofer pupper  puppo  
191                13                  10  None  doggo    None   None  puppo

In [323]:
df_archive[(df_archive.floofer == "floofer") & (df_archive.pupper == "pupper")]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [324]:
df_archive[(df_archive.floofer == "floofer") & (df_archive.puppo == "puppo")]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [325]:
df_archive[(df_archive.pupper == "pupper") & (df_archive.puppo == "puppo")]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

###  Image Predictions: df_img

In [326]:
df_img

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0                                             https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1                                             https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2                                             https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3                                             https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4                                             https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5                                             https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6                                             https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7                                             https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8                                             https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9                                             https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10                                            https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11                                            https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12                                            https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13                                            https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14                                            https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15                                            https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16                                            https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17                                            https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18                                            https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19                                            https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20                                            https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21                                            https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22                                            ht

In [327]:
df_img.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


### Additional Data: df_tweets

In [328]:
df_tweets

created_at                  id              id_str  \
0     Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1     Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2     Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3     Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4     Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   
5     Sat Jul 29 00:08:17 +0000 2017  891087950875897856  891087950875897856   
6     Fri Jul 28 16:27:12 +0000 2017  890971913173991426  890971913173991426   
7     Fri Jul 28 00:22:40 +0000 2017  890729181411237888  890729181411237888   
8     Thu Jul 27 16:25:51 +0000 2017  890609185150312448  890609185150312448   
9     Wed Jul 26 15:59:51 +0000 2017  890240255349198849  890240255349198849   
10    Wed Jul 26 00:31:25 +0000 2017  890006608113172480  890006608113172480   
11    Tue Jul 25 16:11:53 +0000 2017  889880896479866881  889880896479866881   
12    Tue Jul 25 01:55:32 +0000 2017  889665388333682689  889665388333682689   
13    Tue Jul 25 00:10:02 +0000 2017  889638837579907072  889638837579907072   
14    Mon Jul 24 17:02:04 +0000 2017  889531135344209921  889531135344209921   
15    Mon Jul 24 00:19:32 +0000 2017  889278841981685760  889278841981685760   
16    Sun Jul 23 00:22:39 +0000 2017  888917238123831296  888917238123831296   
17    Sat Jul 22 16:56:37 +0000 2017  888804989199671297  888804989199671297   
18    Sat Jul 22 00:23:06 +0000 2017  888554962724278272  888554962724278272   
19    Thu Jul 20 16:49:33 +0000 2017  888078434458587136  888078434458587136   
20    Wed Jul 19 16:06:48 +0000 2017  887705289381826560  887705289381826560   
21    Wed Jul 19 03:39:09 +0000 2017  887517139158093824  887517139158093824   
22    Wed Jul 19 00:47:34 +0000 2017  887473957103951883  887473957103951883   
23    Tue Jul 18 16:08:03 +0000 2017  887343217045368832  887343217045368832   
24    Tue Jul 18 00:07:08 +0000 2017  887101392804085760  887101392804085760   
25    Mon Jul 17 16:17:36 +0000 2017  886983233522544640  886983233522544640   
26    Sun Jul 16 23:58:41 +0000 2017  886736880519319552  886736880519319552   
27    Sun Jul 16 20:14:00 +0000 2017  886680336477933568  886680336477933568   
28    Sat Jul 15 23:25:31 +0000 2017  886366144734445568  886366144734445568   
29    Sat Jul 15 16:51:35 +0000 2017  886267009285017600  886267009285017600   
...                              ...                 ...                 ...   
2316  Tue Nov 17 00:24:19 +0000 2015  666411507551481857  666411507551481857   
2317  Tue Nov 17 00:06:54 +0000 2015  666407126856765440  666407126856765440   
2318  Mon Nov 16 23:23:41 +0000 2015  666396247373291520  666396247373291520   
2319  Mon Nov 16 21:54:18 +0000 2015  666373753744588802  666373753744588802   
2320  Mon Nov 16 21:10:36 +0000 2015  666362758909284353  666362758909284353   
2321  Mon Nov 16 20:32:58 +0000 2015  666353288456101888  666353288456101888   
2322  Mon Nov 16 20:01:42 +0000 2015  666345417576210432  666345417576210432   
2323  Mon Nov 16 19:31:45 +0000 2015  666337882303524864  666337882303524864   
2324  Mon Nov 16 16:37:02 +0000 2015  666293911632134144  666293911632134144   
2325  Mon Nov 16 16:11:11 +0000 2015  666287406224695296  666287406224695296   
2326  Mon Nov 16 15:14:19 +0000 2015  666273097616637952  666273097616637952   
2327  Mon Nov 16 14:57:41 +0000 2015  666268910803644416  666268910803644416   
2328  Mon Nov 16 04:02:55 +0000 2015  666104133288665088  666104133288665088   
2329  Mon Nov 16 03:55:04 +0000 2015  666102155909144576  666102155909144576   
2330  Mon Nov 16 03:44:34 +0000 2015  666099513787052032  666099513787052032   
2331  Mon Nov 16 03:22:39 +0000 2015  666094000022159362  666094000022159362   
2332  Mon Nov 16 02:38:37 +0000 2015  666082916733198337  666082916733198337   
2333  Mon Nov 16 01:59:36 +0000 2015  666073100786774016  666073100786774016   
2334  Mon Nov 16 01:52:02 

In [329]:
df_tweets[df_tweets.is_quote_status == True]

created_at                  id              id_str  \
31    Sat Jul 15 02:45:48 +0000 2017  886054160059072513  886054160059072513   
34    Thu Jul 13 15:19:09 +0000 2017  885518971528720385  885518971528720385   
41    Mon Jul 10 03:08:17 +0000 2017  884247878851493888  884247878851493888   
71    Sat Jun 24 13:24:20 +0000 2017  878604707211726852  878604707211726852   
82    Sun Jun 18 20:30:39 +0000 2017  876537666061221889  876537666061221889   
87    Wed Jun 14 21:06:43 +0000 2017  875097192612077568  875097192612077568   
108   Sat Jun 03 20:33:19 +0000 2017  871102520638267392  871102520638267392   
129   Mon May 22 18:21:28 +0000 2017  866720684873056260  866720684873056260   
185   Sat Apr 22 18:55:51 +0000 2017  855857698524602368  855857698524602368   
187   Sat Apr 22 16:18:34 +0000 2017  855818117272018944  855818117272018944   
236   Mon Mar 27 23:35:28 +0000 2017  846505985330044928  846505985330044928   
238   Sun Mar 26 23:20:02 +0000 2017  846139713627017216  846139713627017216   
251   Tue Mar 21 00:22:10 +0000 2017  843981021012017153  843981021012017153   
263   Mon Mar 13 16:08:50 +0000 2017  841320156043304961  841320156043304961   
277   Tue Mar 07 03:22:35 +0000 2017  838952994649550848  838952994649550848   
316   Tue Feb 21 17:18:39 +0000 2017  834089966724603904  834089966724603904   
327   Fri Feb 17 20:05:43 +0000 2017  832682457690300417  832682457690300417   
396   Tue Jan 24 22:44:42 +0000 2017  824025158776213504  824025158776213504   
450   Sun Jan 08 01:40:55 +0000 2017  817908911860748288  817908911860748288   
531   Wed Dec 07 19:09:37 +0000 2016  806576416489959424  806576416489959424   
543   Fri Dec 02 00:02:45 +0000 2016  804475857670639616  804475857670639616   
568   Tue Nov 22 00:17:10 +0000 2016  800855607700029440  800855607700029440   
611   Tue Nov 08 23:01:49 +0000 2016  796125600683540480  796125600683540480   
667   Fri Oct 21 03:56:25 +0000 2016  789314372632018944  789314372632018944   
797   Sat Sep 03 03:13:29 +0000 2016  771908950375665664  771908950375665664   
883   Sat Jul 30 17:51:13 +0000 2016  759446261539934208  759446261539934208   
1030  Sat Jun 18 17:41:06 +0000 2016  744223424764059648  744223424764059648   
1144  Wed Apr 27 22:57:10 +0000 2016  725458796924002305  725458796924002305   
1190  Sun Apr 03 20:53:33 +0000 2016  716730379797970944  716730379797970944   
1227  Tue Mar 22 16:06:19 +0000 2016  712309440758808576  712309440758808576   
1314  Sat Mar 05 17:26:40 +0000 2016  706169069255446529  706169069255446529   

                                                                                                                                                 text  \
31                                                                                                 RT @Athletics: 12/10 #BATP https://t.co/WxwJmvjfxo   
34                                                                            I have a new hero and his name is Howard. 14/10 https://t.co/gzLHboL7Sk   
41                   OMG HE DIDN'T MEAN TO HE WAS JUST TRYING A LITTLE BARKOUR HE'S SUPER SORRY 13/10 WOULD FORGIVE IMMEDIATE https://t.co/uF3pQ8Wubj   
71                                                                              Martha is stunning how h*ckin dare you. 13/10 https://t.co/9uABQXgjwa   
82       I can say with the pupmost confidence that the doggos who assisted with this search are heroic as h*ck. 14/10 for a… https://t.co/ghfhFSYrMm   
87                           You'll get your package when that precious man is done appreciating the pups. 13/10 for everyone https://t.co/PFp4MghzBW   
108                                                                                                 Never doubt a doggo 14/10 https://t.co/AbBLh2FZCH   
129                                                                        He was providing for his family 13/10 how dare you https://t.co/Q8mVwWN3f4   
185                                 HE'S LIKE "WAIT A MINUTE I'M AN ANIMAL THIS IS AMAZI

In [330]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2346 entries, 0 to 2345
Data columns (total 26 columns):
created_at                       2346 non-null object
id                               2346 non-null int64
id_str                           2346 non-null object
text                             2346 non-null object
truncated                        2346 non-null bool
entities                         2346 non-null object
extended_entities                1824 non-null object
source                           2346 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null object
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null object
in_reply_to_screen_name          77 non-null object
user                             2346 non-null object
geo                              0 non-null object
coordinates                      0 non-null object
place                            1 non-null ob

In [331]:
df_tweets.describe()

id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.346000e+03           7.700000e+01         7.700000e+01   
mean   7.423380e+17           7.440692e+17         2.040329e+16   
std    6.835504e+16           7.524295e+16         1.260797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783824e+17           6.757073e+17         3.589728e+08   
50%    7.189556e+17           7.032559e+17         4.196984e+09   
75%    7.987010e+17           8.233264e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweet_count  favorite_count  
count    2346.000000     2346.000000  
mean     3064.711850     8114.361466  
std      5088.970337    12198.153723  
min         0.000000        0.000000  
25%       612.250000     1410.250000  
50%      1433.500000     3563.500000  
75%      3573.000000    10044.000000  
max     78210.000000   144156.000000

In [332]:
df_tweets[df_tweets.possibly_sensitive == False]

created_at                  id              id_str  \
0     Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1     Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2     Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3     Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4     Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   
5     Sat Jul 29 00:08:17 +0000 2017  891087950875897856  891087950875897856   
6     Fri Jul 28 16:27:12 +0000 2017  890971913173991426  890971913173991426   
7     Fri Jul 28 00:22:40 +0000 2017  890729181411237888  890729181411237888   
8     Thu Jul 27 16:25:51 +0000 2017  890609185150312448  890609185150312448   
9     Wed Jul 26 15:59:51 +0000 2017  890240255349198849  890240255349198849   
10    Wed Jul 26 00:31:25 +0000 2017  890006608113172480  890006608113172480   
11    Tue Jul 25 16:11:53 +0000 2017  889880896479866881  889880896479866881   
12    Tue Jul 25 01:55:32 +0000 2017  889665388333682689  889665388333682689   
13    Tue Jul 25 00:10:02 +0000 2017  889638837579907072  889638837579907072   
14    Mon Jul 24 17:02:04 +0000 2017  889531135344209921  889531135344209921   
15    Mon Jul 24 00:19:32 +0000 2017  889278841981685760  889278841981685760   
16    Sun Jul 23 00:22:39 +0000 2017  888917238123831296  888917238123831296   
17    Sat Jul 22 16:56:37 +0000 2017  888804989199671297  888804989199671297   
18    Sat Jul 22 00:23:06 +0000 2017  888554962724278272  888554962724278272   
19    Thu Jul 20 16:49:33 +0000 2017  888078434458587136  888078434458587136   
20    Wed Jul 19 16:06:48 +0000 2017  887705289381826560  887705289381826560   
21    Wed Jul 19 03:39:09 +0000 2017  887517139158093824  887517139158093824   
22    Wed Jul 19 00:47:34 +0000 2017  887473957103951883  887473957103951883   
23    Tue Jul 18 16:08:03 +0000 2017  887343217045368832  887343217045368832   
24    Tue Jul 18 00:07:08 +0000 2017  887101392804085760  887101392804085760   
25    Mon Jul 17 16:17:36 +0000 2017  886983233522544640  886983233522544640   
26    Sun Jul 16 23:58:41 +0000 2017  886736880519319552  886736880519319552   
27    Sun Jul 16 20:14:00 +0000 2017  886680336477933568  886680336477933568   
28    Sat Jul 15 23:25:31 +0000 2017  886366144734445568  886366144734445568   
30    Sat Jul 15 16:17:19 +0000 2017  886258384151887873  886258384151887873   
...                              ...                 ...                 ...   
2316  Tue Nov 17 00:24:19 +0000 2015  666411507551481857  666411507551481857   
2317  Tue Nov 17 00:06:54 +0000 2015  666407126856765440  666407126856765440   
2318  Mon Nov 16 23:23:41 +0000 2015  666396247373291520  666396247373291520   
2319  Mon Nov 16 21:54:18 +0000 2015  666373753744588802  666373753744588802   
2320  Mon Nov 16 21:10:36 +0000 2015  666362758909284353  666362758909284353   
2321  Mon Nov 16 20:32:58 +0000 2015  666353288456101888  666353288456101888   
2322  Mon Nov 16 20:01:42 +0000 2015  666345417576210432  666345417576210432   
2323  Mon Nov 16 19:31:45 +0000 2015  666337882303524864  666337882303524864   
2324  Mon Nov 16 16:37:02 +0000 2015  666293911632134144  666293911632134144   
2325  Mon Nov 16 16:11:11 +0000 2015  666287406224695296  666287406224695296   
2326  Mon Nov 16 15:14:19 +0000 2015  666273097616637952  666273097616637952   
2327  Mon Nov 16 14:57:41 +0000 2015  666268910803644416  666268910803644416   
2328  Mon Nov 16 04:02:55 +0000 2015  666104133288665088  666104133288665088   
2329  Mon Nov 16 03:55:04 +0000 2015  666102155909144576  666102155909144576   
2330  Mon Nov 16 03:44:34 +0000 2015  666099513787052032  666099513787052032   
2331  Mon Nov 16 03:22:39 +0000 2015  666094000022159362  666094000022159362   
2332  Mon Nov 16 02:38:37 +0000 2015  666082916733198337  666082916733198337   
2333  Mon Nov 16 01:59:36 +0000 2015  666073100786774016  666073100786774016   
2334  Mon Nov 16 01:52:02 

In [333]:
print(df_tweets.favorited.value_counts())
print(df_tweets.retweeted.value_counts())
print(df_tweets.possibly_sensitive.value_counts())

False    2346
Name: favorited, dtype: int64
False    2346
Name: retweeted, dtype: int64
False    2207
Name: possibly_sensitive, dtype: int64


#### Quality

<b>`archive`</b>:
- In_reply_to_status_id has missing data.
- In_reply_to_user_id has missing data.
- Timestamp has +0000 in every cell, and it's a string.
- Some in text column are RT.
- Retweeted_status_id has missing data.
- Retweeted_status_user_id has missing data.
- Retweeted_status_timestamp has missing data.
- Name column: "a", "such" (followed by words "this is"), "None", "quite", "not", etc.
- Wrong ratings:
    - 740373189193256964 should be 14/10
    - 722974582966214656 should be 13/10
    - 716439118184652801 should be 11/10
    - 682962037429899265 should be 10/10
    - 666287406224695296 should be 9/10
    - 786709082849828864 should be 9.75/10
    - 681340665377193984 should be 9.5/10
    
- 835152434251116546 is not a dog being rated. 
- Some rows have more than one stage, even though there is only one dog. <b>This needs to be addressed before tidying the stage columns.</b>
- Some rows have more than one dog. <b>This needs to be addressed before tidying the stage columns.</b>

<b>`img`</b>:
- Some predictions are in lowercase, while some are capitalized.

<b>`tweets`</b>:
- "created at" is a string and has +0000, and format is Day Date Month Time Year, not consistent with "timestamp" in the archive table.


#### Tidiness

<b>`archive`</b>:
- Not needed: 'source', 'retweeted_status_timestamp', 'expanded_urls'.
- The stage names are in individual columns, while they should be values, not variables.

<b>`tweets`</b>:
- Created_at, id, text are duplication of tweet_id, timestamp, and text in `archive` table.
- Not needed: 
    'created_at', 'text','user','id_str', 'truncated', 'entities', 'extended_entities', 'source',
    'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'in_reply_to_screen_name', 
    'geo','coordinates', 'place', 'contributors', 'is_quote_status', 'possibly_sensitive', 
    'possibly_sensitive_appealable','favorited','retweeted', 'lang'
- Tweet id is "id" instead of tweet_id as in other tables.
- `tweets` and `img` tables should be part of `archive` based on id.

# Clean

In [334]:
# Making copies of each table
archive_clean = df_archive.copy()
img_clean = df_img.copy()
tweets_clean = df_tweets.copy()

## Missing Data

#### `archive`
- In_reply_to_status_id has missing data.
- In_reply_to_user_id has missing data.
- Retweeted_status_id has missing data.
- Retweeted_status_user_id has missing data.
- Retweeted_status_timestamp has missing data.


#### Define
Retweets / replies or ones without images are invalid (78 replies and 181 retweets). Drop rows that are retweeted or in response (ones that are non-null) to the original tweets, then remove the columns:
- in_reply_to_status_id and in_reply_to_user_id (78 observations)
- retweeted_status_id and retweeted_status_user_id (181 observations)

#### Code

In [335]:
# If a is not NaN in these columns, it means it's a reply or a retweet, so drop the row
archive_clean = archive_clean.drop(archive_clean[(archive_clean.in_reply_to_status_id.notnull()) | (archive_clean.retweeted_status_id.notnull())].index)

# Removing those columns
archive_clean = archive_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id',
                  'retweeted_status_id', 'retweeted_status_user_id'], axis = 1)

#### Test

In [336]:
archive_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

  retweeted_status_timestamp  \
0                        NaN   
1                        NaN   
2                        NaN   
3                        NaN   
4                        NaN   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0                13                  10   Phineas  None    None   None  None  
1                13                  10     Tilly  None    None   None  None  
2                12                  10    Archie  None    None   None  None  
3                13                  10     Darla  None    None   None  None  
4                12                  10  Franklin  None    None   None  None

<hr>

#### `tweets`
- In_reply_to_status_id has missing data.

#### Define
Retweets are invalid (77 retweets). Drop rows that are in response to the original tweets (ones that are non-null), then remove the columns:
- in_reply_to_status_id and in_reply_to_user_id (77 observations)

#### Code

In [337]:
# If a is not NaN in these columns, it means it's a reply or a retweet, so drop the row
tweets_clean = tweets_clean.drop(tweets_clean[tweets_clean.in_reply_to_status_id.notnull()].index)

# Removing those columns
tweets_clean = tweets_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id'], axis = 1)

#### Test

In [338]:
tweets_clean.head(5)

created_at                  id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1  Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2  Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3  Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4  Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   

                                                                                                                                           text  \
0                                 This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boop… https://t.co/aQFSeaCu9L   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike.… https://t.co/r0YlrsGCgy   
3                                       This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected a… https://t.co/0g0KMIVXZ3   

   truncated  \
0      False   
1       True   
2       True   
3      False   
4       True   

                                                                                                                                                                                                  entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media...   
1  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/aQFSeaCu9L', 'expanded_url': 'https://twitter.com/i/web/status/892177421306343426', 'display_url': 'twitter.com/...   
2  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/r0YlrsGCgy', 'expanded_url': 'https://twitter.com/i/web/status/891815181378084864', 'display_url': 'twitter.com/...   
3  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 891689552724799489, 'id_str': '891689552724799489', 'indices': [80, 103], 'media_url': 'http://pbs.twimg.com/media...   
4  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/0g0KMIVXZ3', 'expanded_url': 'https://twitter.com/i/web/status/891327558926688256', 'display_url': 'twitter.com/...   

                                                                                                                                                                                         extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com...   
1                                                                                                                                                                                                      NaN   
2                                                                                                                                                                                                      NaN   
3  {'media': [{'id': 891689552724799489, 'id_str': '891689552724799489', 'indices': [80, 103], 'media_url': 'http://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg', 'media_url_https': 'https://pbs.twimg.com...   
4                                                                                                                                                                                                      NaN   

                                                                               source  \
0  <a href="http:

## Tidiness

`archive`
- Not needed: 'source', 'retweeted_status_timestamp', 'expanded_urls'.

`tweets`
- Created_at, id, text are duplication of tweet_id, timestamp, and text in `archive` table.
- Not needed: 
    'created_at', 'text','user','id_str', 'truncated', 'entities', 'extended_entities', 'source',
    'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'in_reply_to_screen_name', 
    'geo','coordinates', 'place', 'contributors', 'is_quote_status', 'possibly_sensitive', 
    'possibly_sensitive_appealable','favorited','retweeted', 'lang'


#### Define
Remove the "not-needed" columns from the `archive` and `tweets` tables, as well as the "created_at" and "text" columns from `tweets` table. Keep "id" column to merge the tables later.

#### Code

In [339]:
archive_drop = ['source', 'retweeted_status_timestamp', 'expanded_urls']
tweets_drop = ['created_at', 'text','user','id_str', 'truncated', 'entities', 'extended_entities', 'source',
    'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'in_reply_to_screen_name', 
    'geo','coordinates', 'place', 'contributors', 'is_quote_status', 'possibly_sensitive', 
    'possibly_sensitive_appealable','favorited','retweeted', 'lang']

# Dropping columns I don't need
archive_clean = archive_clean.drop(archive_drop, axis=1) 
tweets_clean = tweets_clean.drop(tweets_drop, axis=1) 

#### Test

In [340]:
archive_clean.head(3)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None  
2                12                  10   Archie  None    None   None  None

In [341]:
tweets_clean.head(3)

id  retweet_count  favorite_count
0  892420643555336193           8694           39097
1  892177421306343426           6383           33488
2  891815181378084864           4241           25237

<hr>

`tweets` 

- id is "id" instead of tweet_id as in other tables

#### Define

Change column name from 'id' to 'tweet_id'

#### Code

In [342]:
tweets_clean.columns = ['tweet_id','retweet_count','favorite_count']

#### Test

In [343]:
tweets_clean.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8694           39097
1  892177421306343426           6383           33488
2  891815181378084864           4241           25237
3  891689557279858688           8794           42457
4  891327558926688256           9576           40620

<hr>

`tweets` `img`

- `tweets` and `img` tables should be part of `archive` based on id

#### Define

Merge all tables, joining on tweet_id in the `img` table.

#### Code

In [344]:
archive_tweets = pd.merge(archive_clean, tweets_clean, on='tweet_id', how='inner')
master = pd.merge(archive_tweets, img_clean, on='tweet_id', how='inner')

#### Test

In [345]:
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1971 entries, 0 to 1970
Data columns (total 23 columns):
tweet_id              1971 non-null int64
timestamp             1971 non-null object
text                  1971 non-null object
rating_numerator      1971 non-null int64
rating_denominator    1971 non-null int64
name                  1971 non-null object
doggo                 1971 non-null object
floofer               1971 non-null object
pupper                1971 non-null object
puppo                 1971 non-null object
retweet_count         1971 non-null int64
favorite_count        1971 non-null int64
jpg_url               1971 non-null object
img_num               1971 non-null int64
p1                    1971 non-null object
p1_conf               1971 non-null float64
p1_dog                1971 non-null bool
p2                    1971 non-null object
p2_conf               1971 non-null float64
p2_dog                1971 non-null bool
p3                    1971 non-null objec

<hr>

`archive`

- Some rows have more than one stage, even though there is only one dog. <b>This needs to be addressed before tidying the stage columns.</b>
- Some rows have more than one dog. <b>This needs to be addressed before tidying the stage columns.</b>
- The stage names are values, not variables

#### Define

1. Check which rows have this issue, and correct them with more than one stage by reading the text.
- Multiple stages are all 'doggo' and 'pupper', so make a new value called "doggo_pupper" inside the 'doggo' column, and make 'puppo' column 'None'.
- Then make a column called "stage" and fill the values. 
- Then drop the original columns.

In [346]:
master[(master.doggo == "doggo") & (master.floofer == "floofer")]

tweet_id                  timestamp  \
152  854010172552949760  2017-04-17 16:34:26 +0000   

                                                                                                                                                                     text  \
152  At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk   

     rating_numerator  rating_denominator  name  doggo  floofer pupper puppo  \
152                11                  10  None  doggo  floofer   None  None   

      ...    img_num                p1   p1_conf  p1_dog      p2   p2_conf  \
152   ...          1  English_springer  0.354733    True  collie  0.177538   

     p2_dog             p3   p3_conf  p3_dog  
152    True  Border_collie  0.131706    True  

[1 rows x 23 columns]

In [347]:
master[(master.doggo == "doggo") & (master.pupper == "pupper")]

tweet_id                  timestamp  \
337  817777686764523521  2017-01-07 16:59:28 +0000   
394  808106460588765185  2016-12-12 00:29:28 +0000   
421  801115127852503040  2016-11-22 17:28:25 +0000   
506  785639753186217984  2016-10-11 00:34:48 +0000   
648  759793422261743616  2016-07-31 16:50:42 +0000   
700  751583847268179968  2016-07-09 01:08:47 +0000   
789  741067306818797568  2016-06-10 00:39:48 +0000   
835  733109485275860992  2016-05-19 01:38:16 +0000   

                                                                                                                                                                     text  \
337  This is Dido. She's playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7   
394                      Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   
421                      This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj   
506  This is Pinot. He's a sophisticated doggo. You can tell by the hat. Also pointier than your average pupper. Still 10/10 would pet cautiously https://t.co/f2wmLZTPHd   
648                      Meet Maggie &amp; Lila. Maggie is the doggo, Lila is the pupper. They are sisters. Both 12/10 would pet at the same time https://t.co/MYwR4DQKll   
700                              Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8   
789                                                                           This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC   
835                                                                                            Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda   

     rating_numerator  rating_denominator    name  doggo floofer  pupper  \
337                13                  10    Dido  doggo    None  pupper   
394                12                  10    None  doggo    None  pupper   
421                12                  10   Bones  doggo    None  pupper   
506                10                  10   Pinot  doggo    None  pupper   
648                12                  10  Maggie  doggo    None  pupper   
700                 5                  10    None  doggo    None  pupper   
789                12                  10    just  doggo    None  pupper   
835                12                  10    None  doggo    None  pupper   

    puppo   ...    img_num                      p1   p1_conf  p1_dog  \
337  None   ...          1  curly-coated_retriever  0.733256    True   
394  None   ...          1        golden_retriever  0.426183    True   
421  None   ...          1               dalmatian  0.823356    True   
506  None   ...          1               porcupine  0.978042   False   
648  None   ...          2        golden_retriever  0.985876    True   
700  None   ...          1               dalmatian  0.868304    True   
789  None   ...          1        golden_retriever  0.843799    True   
835  None   ...          1        golden_retriever  0.945523    True   

                        p2   p2_conf  p2_dog                   p3   p3_conf  \
337  flat-coated_retriever  0.214145    True  Irish_water_spaniel  0.029769   
394     Labrador_retriever  0.257447    True       Great_Pyrenees  0.126482   
421         English_setter  0.094602    True             bluetick  0.021953   
506             sea_urchin  0.006106   False              echidna  0.005442   
648     Labrador_retriever  0.001948    True               kuvasz  0.001752   
700           studio_couch  0.059623   False         snow_leopard  0.013876   
789     Labrador_retriever  0.052956    True               kelpie  0.035711   
835 

In [348]:
master[(master.doggo == "doggo") & (master.puppo == "puppo")]

tweet_id                  timestamp  \
146  855851453814013952  2017-04-22 18:31:02 +0000   

                                                                                                                                                               text  \
146  Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel   

     rating_numerator  rating_denominator  name  doggo floofer pupper  puppo  \
146                13                  10  None  doggo    None   None  puppo   

      ...    img_num                     p1   p1_conf  p1_dog  \
146   ...          1  flat-coated_retriever  0.321676    True   

                     p2   p2_conf  p2_dog           p3  p3_conf  p3_dog  
146  Labrador_retriever  0.115138    True  groenendael   0.0961    True  

[1 rows x 23 columns]

#### Code

In [349]:
# Correcting the rows with two stages when there should be only one
master.loc[master.tweet_id == 854010172552949760, 'floofer'] = 'None'
master.loc[master.tweet_id == 817777686764523521, 'pupper'] = 'None'
master.loc[master.tweet_id == 801115127852503040, 'doggo'] = 'None'
master.loc[master.tweet_id == 785639753186217984, 'pupper'] = 'None'
master.loc[master.tweet_id == 751583847268179968, 'pupper'] = 'None'
master.loc[master.tweet_id == 855851453814013952, 'doggo'] = 'None'

# Changing the two-dog rows to "doggo_pupper" and replacing "pupper" with "None".
doggo_pupper_array = [808106460588765185, 759793422261743616, 741067306818797568, 733109485275860992]
def doggo_pupper(id):
    master.loc[master.tweet_id == id, 'doggo'] = 'doggo_pupper'
    master.loc[master.tweet_id == id, 'pupper'] = 'None'

for id in doggo_pupper_array:
    doggo_pupper(id)

In [350]:
master.loc[master.tweet_id == 808106460588765185]

tweet_id                  timestamp  \
394  808106460588765185  2016-12-12 00:29:28 +0000   

                                                                                                                                                 text  \
394  Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   

     rating_numerator  rating_denominator  name         doggo floofer pupper  \
394                12                  10  None  doggo_pupper    None   None   

    puppo   ...    img_num                p1   p1_conf  p1_dog  \
394  None   ...          1  golden_retriever  0.426183    True   

                     p2   p2_conf  p2_dog              p3   p3_conf  p3_dog  
394  Labrador_retriever  0.257447    True  Great_Pyrenees  0.126482    True  

[1 rows x 23 columns]

In [351]:
# Reference: https://stackoverflow.com/questions/19913659/pandas-conditional-creation-of-a-series-dataframe-column
conditions = [
    (master['doggo'] == 'doggo'),
    (master['doggo'] == 'doggo_pupper'),
    (master['floofer'] == 'floofer'),
    (master['pupper'] == 'pupper'),
    (master['puppo'] == 'puppo')]
choices = ['doggo', 'doggo_pupper', 'floofer', 'pupper', 'puppo']
master['stage'] = np.select(conditions, choices, default='unknown')

# Dropping the individual stage columns
master = master.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1)

#### Test

In [352]:
master.stage.value_counts()

unknown         1668
pupper           202
doggo             67
puppo             23
floofer            7
doggo_pupper       4
Name: stage, dtype: int64

## Quality

`archive`
- Timestamp has +0000 in every cell, and it's a string.

#### Define

Using to_datetime from pandas to make the dtype datetime and remove +0000

#### Code

In [353]:
master.timestamp = pd.to_datetime(master.timestamp)

#### Test

In [354]:
master.dtypes

tweet_id                       int64
timestamp             datetime64[ns]
text                          object
rating_numerator               int64
rating_denominator             int64
name                          object
retweet_count                  int64
favorite_count                 int64
jpg_url                       object
img_num                        int64
p1                            object
p1_conf                      float64
p1_dog                          bool
p2                            object
p2_conf                      float64
p2_dog                          bool
p3                            object
p3_conf                      float64
p3_dog                          bool
stage                         object
dtype: object

<hr>

`archive`
- Name column: "a", "such" (followed by words "this is"), "None", "quite", "not", etc.

#### Define

The names seemed to have been extracted from the word "this is", and incorrect dog names are in lowercase. I can find the lowercase names and replace them with "None". 

#### Code

In [355]:
# Subsetting the rows with incorrect names
wrong_name = master.name.str.islower()

# Replacing them with 'None'
master.loc[wrong_name, 'name'] = 'None'

#### Test

In [356]:
master[master.name.str.islower() == True]

Empty DataFrame
Columns: [tweet_id, timestamp, text, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, stage]
Index: []

<hr>

`archive`
- Wrong ratings:
    - 740373189193256964 should be 14/10
    - 722974582966214656 should be 13/10
    - 716439118184652801 should be 11/10
    - 682962037429899265 should be 10/10
    - 666287406224695296 should be 9/10
    - 786709082849828864 should be 9.75/10
    - 681340665377193984 should be 9.5/10
    
- 835152434251116546 is not a dog being rated.

#### Define

Locate the tweet_ids and replace with correct numerators and denominators.
Remove 835152434251116546. 

#### Code

In [357]:
# Function to locate a tweet_id and change its numerator and denominator
def change_rating(id, num, den):
    master.loc[master.tweet_id == id, 'rating_numerator'] = num
    master.loc[master.tweet_id == id, 'rating_denominator'] = den

change_rating(740373189193256964, 14, 10)
change_rating(722974582966214656, 13, 10)
change_rating(716439118184652801, 11, 10)
change_rating(682962037429899265, 10, 10)
change_rating(666287406224695296, 9, 10)
change_rating(786709082849828864, 9.75, 10)
change_rating(681340665377193984, 9.5, 10)

# Remove this row
master[master.tweet_id == 835152434251116546]
master = master.drop(master.index[[232]])


#### Test

In [358]:
master[(master.tweet_id == 740373189193256964) |
       (master.tweet_id == 722974582966214656) |
       (master.tweet_id == 716439118184652801) |
       (master.tweet_id == 682962037429899265) |
       (master.tweet_id == 666287406224695296) |
        (master.tweet_id == 786709082849828864) |
      (master.tweet_id == 681340665377193984)]

tweet_id           timestamp  \
499   786709082849828864 2016-10-13 23:23:56   
794   740373189193256964 2016-06-08 02:41:38   
884   722974582966214656 2016-04-21 02:25:47   
918   716439118184652801 2016-04-03 01:36:11   
1313  682962037429899265 2016-01-01 16:30:13   
1950  666287406224695296 2015-11-16 16:11:11   

                                                                                                                                              text  \
499               This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   
794   After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
884                                                                               Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
918                          This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1313   This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   
1950      This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

      rating_numerator  rating_denominator      name  retweet_count  \
499               9.75                  10     Logan           6930   
794              14.00                  10      None          14798   
884              13.00                  10      None           1730   
918              11.00                  10  Bluebert            243   
1313             10.00                  10    Darrel          18667   
1950              9.00                  10      None             68   

      favorite_count                                          jpg_url  \
499            20025  https://pbs.twimg.com/media/CurzvFTXgAA2_AP.jpg   
794            37468  https://pbs.twimg.com/media/CkZVdJ6WYAAXZ5A.jpg   
884             4421  https://pbs.twimg.com/media/CgiFjIpWgAA4wVp.jpg   
918             2536  https://pbs.twimg.com/media/CfFNk7cWAAA-hND.jpg   
1313           40060  https://pbs.twimg.com/media/CXpeVzQW8AApKYb.jpg   
1950             151  https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   

      img_num                p1   p1_conf  p1_dog                          p2  \
499         1        Pomeranian  0.467321    True                 Persian_cat   
794         3  golden_retriever  0.807644    True                      kuvasz   
884         1        Great_Dane  0.246762    True  Greater_Swiss_Mountain_dog   
918         1    Siberian_husky  0.396495    True                    malamute   
1313        1             dingo  0.278600   False                   Chihuahua   
1950        1       Maltese_dog  0.857531    True                  toy_poodle   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog    stage  
499   0.122978   False                chow  0.102654    True  unknown  
794   0.101286    True  Labrador_retriever  0.023785    True  unknown  
884   0.126131    True          Weimaraner  0.085297    True  unknown  
918   0.317053    True          Eskimo_dog  0.273419    True  unknown  
1313  0.155207    True               loupe  0.153598   False  unknown  
1950  0.063064    True    miniature_poodle  0.025581    True  unknown

In [359]:
master[master.tweet_id == 835152434251116546]

Empty DataFrame
Columns: [tweet_id, timestamp, text, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, stage]
Index: []

<hr>

`img`
- Some predictions are in lowercase, while some are capitalized.

#### Define

For consistency, make everything lowercase

#### Code

In [360]:
master.p1 = master.p1.str.lower()
master.p2 = master.p2.str.lower()
master.p3 = master.p3.str.lower()

#### Test

In [361]:
master.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   rating_numerator  rating_denominator      name  retweet_count  \
0              13.0                  10   Phineas           8694   
1              13.0                  10     Tilly           6383   
2              12.0                  10    Archie           4241   
3              13.0                  10     Darla           8794   
4              12.0                  10  Franklin           9576   

   favorite_count                                          jpg_url  img_num  \
0           39097  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1   
1           33488  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1   
2           25237  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1   
3           42457  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1   
4           40620  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2   

            p1   p1_conf  p1_dog                  p2   p2_conf  p2_dog  \
0       orange  0.097049   False               bagel  0.085851   False   
1    chihuahua  0.323581    True            pekinese  0.090647    True   
2    chihuahua  0.716012    True            malamute  0.078253    True   
3  paper_towel  0.170278   False  labrador_retriever  0.168086    True   
4       basset  0.555712    True    english_springer  0.225770    True   

                            p3   p3_conf  p3_dog    stage  
0                       banana  0.076110   False  unknown  
1                     papillon  0.068957    True  unknown  
2                       kelpie  0.031379    True  unknown  
3                      spatula  0.040836   False  unknown  
4  german_short-haired_pointer  0.175219    True  unknown

In [362]:
master.to_csv('twitter_archive_master.csv')